Update release notes 0.7 (#560)
harness/developer-hub: release-notes/chaos-engineering.md

Additionally, the release notes below are only for NextGen SaaS. FirstGen SaaS release notes are available [here](/docs/first-gen/firstgen-release-notes/harness-saa-s-release-notes) and Self-Managed Enterprise Edition release notes are available [here](/release-notes/self-managed-enterprise-edition).
:::

## January 17, 2023, version 0.7

### What's new

* Resilience tab introduced on the pipeline execution screen (Chaos-963)
    * This release adds a resilience tab, which displays the experiment results as a list of probes, their logs (descriptions), and probe status. Instead of navigating to the **View detailed execution** section of the experiment, you can now view the results of all the experiments on the pipeline execution screen. 

* Support for X-API-KEY authentication (Chaos-667)

>     * This release adds support for X-API-KEY authentication for user facing chaos APIs. This way, you can avoid using the JWT token which gives more control over the module and set your own custom expiration time on the X-API-KEY. 

Author: No, the JWT token gives more access.

> * Supports deployment on existing chaos infrastructure

Author: Sure, got it, Deepak.

>     * This release adds support for deploying your project on existing chaos infrastructure (also known as the greenfield method of deployment). You can specify the name of the delegate and other details like installation mode, service account name, and namespace, after which the YAML manifest is generated and sent over to the cluster rather than being downloaded on your system. Once the cluster receives the manifest, the delegate deploys your project on the selected infrastructure. Currently, you can only connect to the chaos infrastructure using account-level delegates.

Reviewer: ```suggestion
    * This release adds support for deploying your chaos infrastructures on cluster via existing deployed Harness delegates (also known as the brownfield method of deployment). You can select the connector pointing to the required delegate and other details like installation mode, service account name, and namespace, after which the YAML manifest is generated and sent over to the cluster rather than being downloaded on your system. Once the delegate receives the manifest, It deploys your chaos infrastructure on the selected cluster. Currently, you can only deploy chaos infrastructure using account-level delegates.
```

* Details of an experiment are pre-filled when adding it to chaoshub (Chaos-989)
    * Instead of having to re-enter the same details, this release pre-fills the details of the experiment that you want to add to chaoshub. You can simply navigate to the experiment, and click on **Add to ChaosHub**. This will display a screen with the name of the experiment, a description (optional), and tags (optional). You can add your experiment to the chaoshub of your choice by clicking **Save**.

* One sync retry to connect to a disconnected chaoshub (Chaos-999)

>     * When you try to connect to a chaoshub that is disconnected, you will not see any faults or experiments in the chaoshub. This release adds a feature such that when you click on a disconnected chaoshub, it tries to synchronize and connect to the chaoshub at least once. 

Reviewer: ```suggestion
    * When you try to open a chaoshub that is disconnected, you will not see any faults or experiments in the chaoshub. This release adds a feature such that when you click on a disconnected chaoshub, it tries to synchronize and connect to the chaoshub at least once. 
```

* Filter chaos experiment based on target infrastructure (Chaos-959)

>     * In the **Deployment** tab, when you click on **Pipeline** and select a chaos experiment, you can filter experiments based on their names. This release adds another filter so that you can view experiments based on the target infrastructure.

Reviewer: ```suggestion
    * In the **Deployment** tab, when you click on **Pipeline** and select a chaos experiment, you can filter experiments based on their names. This release adds another filter so that you can view experiments based on the target chaos infrastructure.
```

    * When a user enters an incorrect URL in their browser when they are inside a chaos experiment, previously the user interface would show a blank screen. This release displays an error message on the user interface stating that the entered URL is invalid.

* Last updated by field shows updated user name (Chaos-916)
    * In a chaos experiment, when an experiment is saved, it shows the **Resilience score** field, **Last run status**, and **Last updated by** as **N/A**. This release updates the  **Last updated by** field with the name of the user who updated the chaos experiment most recently.

* Average resilience score shows difference between current and last executed resilience scores (Chaos-916)
    * In the chaos experiments tab, the **Resilience score** field displays the resilience score as well as the percentage increase in resilience scores between the current and previous runs. This release removes the percentage increase, and instead, displays the difference between the current run and previous run’s resilience score for the same experiment for better readability.

* Experiments overview page categorizes experiments based on average resilience score (Chaos-802)

>     * This release categorizes and displays all the chaos experiments based on the average resilience score. It also displays the number of experiments in each category. It shows three categories, i.e., experiments with an average resilience score between 0 and 40, 40 to 80, and 80 to 100. This provides better insights about the chaos experiments and their resilience scores. Previously, the overview page only showed the number of experiments that passed and the number of experiments that failed. 

Reviewer 2: ```suggestion
    * This release categorizes and displays all the chaos experiments on the basis of average resilience score. It also displays the number of experiments in each category. It shows three categories, which are for experiments with an average resilience score between 0 and 40, 40 to 80, and 80 to 100. This provides better insights about the chaos experiments and their resilience scores. Previously, the overview page showed only the number of experiments that passed and the number of experiments that failed. 
```
* Check whether I use the right words to replace "i.e.". 
* The first category was described as "between 0 and 40. That wording excludes the values at the extremes (0 and 40). So does the resilience score begin with 1? Or is zero included? If it is included, try "from 0 through 40". However, in that case, 40 lies in two categories. If the extremes are to be included in all categories, even 80 falls in two categories. If an experiment has a resilience score of exactly 80, in which category is it shown? Should we write "from 0 through 39, 40 through 79, and 80 through 100"? That's an example. We will have to find out from engineering.
* Also, global comment: in some descriptions, we describe previous behavior first. Here (and I noticed it elsewhere), we describe the previous behavior at the end. This is not a critical issue, but it helps to be consistent in how we organize our information for easy parsing.

* Every run of an experiment is clickable to view detailed execution (Chaos-1032)

>     * In the **Chaos Experiments** tab, you can see the detailed execution of an experiment's runs by clicking on **View run**. This release makes every run of an experiment clickable so that you can view the detailed execution of an experiment by just clicking on the experiment.

Reviewer: ```suggestion
    * In the **Chaos Experiments** tab, you can see the detailed execution of an experiment's runs by clicking on **View run**. This release makes every run of an experiment clickable so that you can view the detailed execution of an experiment run by just clicking on the experiment run row.
```

>     * In the **Chaos Experiments** tab, you can see the detailed execution of an experiment's runs by clicking on **View run**. This release makes every run of an experiment clickable so that you can view the detailed execution of an experiment run by just clicking on the experiment run row.

Reviewer 2: ```suggestion
    * In the **Chaos Experiments** tab, you can see the detailed execution of an experiment's runs by clicking on **View run**. This release makes every run of an experiment clickable so that you can view the detailed execution of an experiment run.
```
Here's my interpretation of this description: both releases (previous and new) showed detailed information about every run. The previous release did that by providing a button named **View run** next to every run (or maybe a single button that expanded all runs at once). The current release does that by removing the button and making the run (I take "run" here to mean whatever text is used to show a run) itself clickable. Is my interpretation correct? If yes, maybe we can be a little more precise. I re-read this description multiple times before I reached this conclusion. If this change is merely about the removal of a button and addition of equivalent functionality, we should say so. Let me know whether I'm right. Then we can discuss rewording if necessary.

### Early access
* There are no early access features in this release. 

### Fixed issues

> * Searching for chaos experiments in the search bar only showed experiments that had been run at least once. Now it has been fixed such that when you search for an experiment, the search results include those experiments that were aborted and experiments that were saved but not run. (Chaos 916)

Reviewer 2: ```suggestion
* Searching for chaos experiments by using the search bar showed only those experiments that had been run at least once. Now, when you search for an experiment, the search results include those experiments that were aborted and experiments that were saved but not run. (Chaos 916)
```
(Deleted text that would have made it necessary to be clear on what "it" refers to (namely, the search functionality))

> * If some of the parameters were left as empty strings when specifying the target application parameters through a YAML manifest, the user interface for the target application page would crash. This has been fixed so that irrespective of the values entered in the YAML manifest, you can change the values of the target application on the user interface. (Chaos-970)

Reviewer 2: ```suggestion
* When specifying the target application parameters through a YAML manifest, if you left some parameters empty, the user interface of the target application page would stop responding. This has been fixed so that, irrespective of the values that you enter in the YAML manifest, you can change the values of the target application on the user interface. (Chaos-970)
```

> * In a chaos experiment, the fault library would incorrectly show fault categories and fault labels even when the hub had been disconnected. It would persist data from the previously selected chaos hub. This has now been fixed. When you click on a disconnected chaos hub, it displays the message “No faults found in the selected hub”. (Chaos 971)

Reviewer 2: ```suggestion
* In a chaos experiment, the fault library would incorrectly show fault categories and fault labels even when the hub had been disconnected. The fault library would persist data from the previously selected chaos hub. This has now been fixed. A disconnected chaos hub now displays the message “No faults found in the selected hub.” (Chaos-971)
```

> * On the chaos hub screen, you could not scroll over the list of hubs from all places on the screen. Now the scroll bar has been shifted to the extreme right so that you can scroll through from anywhere on the screen. (Chaos-964)

Reviewer 2: ```suggestion
* On the chaos hub screen, you could not scroll through the list of hubs from any location on the screen. This issue has been fixed by moving the scroll bar to the extreme right of the screen. (Chaos-964)
```
(reworded the second sentence to avoid repeating a number of words)

> * When you hovered over a probe, the name of the probe would not be displayed fully if it was long. Now it has been fixed. (Chaos-990)

Reviewer: ```suggestion
* When you hovered over a probe, the details of the probe would overflow if it was long. Now it has been fixed. (Chaos-990)
```

> * If you hovered over a probe, its details would overflow if it was long. Now it has been fixed. (Chaos-990)

Reviewer 2: ```suggestion
* If you hovered over a probe, its details would overflow if they were too long. Now, it has been fixed. (Chaos-990)
```

> * In the chaos hub, when the number of chaos faults that you wished to view on a page was altered (or increased), it would result in a blank page. Now it has been fixed. (Chaos-984)

Reviewer 2: Who altered the list? The user? Good to write in the active voice if you know who the actor is (or can explicitly name them). 
```suggestion
* In a chaos hub, if you altered the number of chaos faults that you wished to view on a page, it would result in a blank page. Now it has been fixed. (Chaos-984)
```
Again, we should try to avoid the weak (or empty) "it" as far as possible. How about "Any alteration to the number of chaos faults that you wished to view on a single page in a chaos hub would result in a blank page." (Please reword if necessary.)
Also, is it any alteration or only an increase? Be specific. Don't write both.

> * When a chaos experiment was imported into the chaos hub, it would not be logged as an audit event and would not be displayed on the user interface. It has been fixed. (Chaos-779)

Reviewer 2: What would not be displayed on the UI? The chaos experiment itself or the audit event entry? Currently, the wording suggests that it could be either. 
Also, optional to reword: "The issue is now fixed" is better than "It has been fixed." (The latter is not wrong at all; it just makes us look like we are in a hurry to finish the description.)

> * If no chaos infrastructure is connected with your project, it previously showed a blank screen. Now it displays the message, "There are no chaos infrastructure in your project." (Chaos-1009) 

Reviewer 2: The empty "it" resurfaces. What showed a blank screen? Again in the second sentence. Also, the message needs a grammatical correction. Please let engineering know. Thanks.

> * In CRON experiments, the scheduled run time would always be shown in GMT. Now it has been fixed to show the run time in your browser’s time zone. (Chaos-1035)

Reviewer 2: ```suggestion
* In CRON experiments, the scheduled run time would always be shown in GMT. Now, it has been fixed to show the run time in your browser’s time zone. (Chaos-1035)
```

> * The parameters in the YAML manifest of different runs of the same chaos experiment were inconsistent with the changes made (if any) in their respective runs. Now it has been fixed.

Reviewer 2: ```suggestion
* The parameters in the YAML manifest of different runs of the same chaos experiment were inconsistent with the changes made (if any) in their respective runs. Now, it has been fixed.
```

    * If you enter an incorrect URL in your browser when viewing a chaos experiment, previously, the user interface would show a blank screen. This release displays an error message stating that the entered URL is invalid.

* The **Last updated by** field shows a user name (CHAOS-916)
    * A saved chaos experiment shows the **Resilience score**, **Last run status**, and **Last updated by** fields as **N/A**. This release updates the  **Last updated by** field with the name of the user who updated the chaos experiment most recently.

* Average resilience score shows the difference between the current and last executed resilience scores (CHAOS-916)
    * On the chaos experiments tab, the **Resilience score** field displayed the resilience score and the percentage increase in resilience score between the current and previous runs of an experiment. This release removes the percentage increase and, instead, displays the difference between the current run's and previous run’s resilience score for better readability.

* The **Experiments overview** page categorizes experiments on the basis of average resilience score (CHAOS-802)

>     * This release categorizes and displays all the chaos experiments on the basis of average resilience score. It also displays the number of experiments in each category. It shows three categories, which corresponds to experiments with an average resilience score between 0 through 39, 40 through 79, and 80 through 100. This provides better insights about the chaos experiments and their resilience scores. Previously, the overview page showed only the number of experiments that passed and the number of experiments that failed. 

Reviewer 3: ```suggestion
    * This release categorizes and displays all the chaos experiments on the basis of average resilience score. It also displays the number of experiments in each category. It shows three categories based on the average resilience scores of the experiments: 0 through 39, 40 through 79, and 80 through 100. This provides better insights about the chaos experiments and their resilience scores. Previously, the overview page showed only the number of experiments that passed and the number of experiments that failed. 
```

* In a chaos experiment, the fault library would incorrectly show fault categories and fault labels even when the hub had been disconnected. The fault library would persist data from the previously selected ChaosHub. This has now been fixed. A disconnected ChaosHub now displays the message “No faults found in the selected hub.” (CHAOS-971)
* On the ChaosHub screen, you could not scroll through the list of hubs from any location on the screen. This issue has been fixed by moving the scroll bar to the extreme right of the screen. (CHAOS-964)
* If you hovered over a probe, its details would overflow if they were too long. Now, it has been fixed. (CHAOS-990)
* Any increase in the number of chaos faults that you wished to view on a single page in a ChaosHub would result in a blank page. Now, it has been fixed. (CHAOS-984)
* When a chaos experiment was imported into the ChaosHub, it would not be logged as an audit event and would not be displayed on the user interface. It has been fixed. (CHAOS-779)

> * If no chaos infrastructure is connected with your project, a blank screen would be displayed. Now, the message "There are no chaos infrastructure in your project." is displayed. (CHAOS-1009) 

Reviewer 3: "There are no chaos infrastructure in your project." Is this literally what the UI says? Seems to me it should say "There is no chaos infrastructure" (single) or "There are no chaose infrastructures" (plural).

Author: Oh yes, the statement is grammatically incorrect. I will change it here. I have already raised a ticket for this with the engg team.

* In CRON experiments, the scheduled run time would always be shown in GMT. Now, it has been fixed to show the run time in your browser’s time zone. (CHAOS-1035)

> * The parameters in the YAML manifest of different runs of the same chaos experiment were inconsistent with the changes made (if any) in their respective runs. Now, it has been fixed.

Reviewer 3: Missing the Jira ticket number.

## December 23, 2022, version 0.6

### What’s new

* Optimized listWorkflow and listWorkflowRun queries in the chaos manager (CHAOS-860)
    * This release optimizes the listWorkflow and listWorkflowRun queries in the chaos manager by only fetching those experiments that the user requests, instead of loading all the experiments at once.

* Pagination on the faults and experiments screen (CHAOS-689)
    * This release adds pagination on the faults and experiment screen in ChaosHub that allows you to scroll and navigate through the experiments by pages. 

> * Enable save and run buttons on the experiment builder (CHAOS-913)

Reviewer 3: **Save** and **Run**? Capitalized and bolded?

>     * This release enables the save and run buttons after you tune the application by specifying the parameters on the user interface. As a consequence, the default weight is set to 10 since the user would not move to the next step of setting fault weights.

Reviewer 3: Same as previous.

 
* Experiment can be viewed during execution (CHAOS-835)
    * This release allows you to view the experiment even when it is being executed. Previously, an experiment could be viewed only after the run was complete.  

* Edit chaos experiment is separated into two action components (CHAOS-685)

>     * This release divides the edit experiment action into two actions: edit experiment and clone experiment. The edit experiment action helps you make changes to the current (or selected) experiment. The clone experiment action helps the user create a new experiment from an already existing experiment. The cloned experiment retains the same configuration as the original experiment with the ability to tune the configurations if required.

Reviewer 3: ```suggestion
    * This release divides the edit experiment action into two actions: Edit Experiment and Clone Experiment. The Edit Experiment action helps you make changes to the current (or selected) experiment. The Clone Experiment action helps you create a new experiment from an already existing experiment. The cloned experiment retains the same configuration as the original experiment with the ability to tune the configurations if required.
```

* Experiments executed and triggered by respective categories (a pipeline, a scheduled CRON job, or a user) are correctly shown. (CHAOS-800)
* When a chaos experiment contains characters such as ‘ ‘, ‘/’, and so on, logs are correctly parsed and displayed on the screen. The execution is encoded before being sent to the control plane and decoded after being received by the user interface. (CHAOS-854)
* After deleting a chaos experiment from a particular page, the pagination is reset and only shows the available experiments. (CHAOS-923)
* When a chaos infrastructure is deleted, details on the user interface wrongly showed the infrastructure ID instead of the infrastructure name. This is now fixed. (CHAOS-952)
* When a chaos experiment was pushed to the ChaosHub, only a single fault associated with the experiment was being pushed, rather than all the faults. This is now fixed. (CHAOS-973)
* When a chaos experiment was deleted, only the most recent run was deleted, and the previous runs were retained in the cluster. Now it has been fixed such that when a chaos experiment is deleted, all the runs associated with it are deleted from the cluster. 

> * When a chaos experiment was deleted, the fault running within the experiment was not stopped). Now it has been fixed such that, when an experiment is deleted, the chaos fault running on the Kubernetes cluster is halted, the fault is deleted, and the experiment (as a whole) is deleted. (CHAOS-782)

Reviewer 3: ```suggestion
* When a chaos experiment was deleted, the fault running within the experiment was not stopped. Now it has been fixed such that, when an experiment is deleted, the chaos fault running on the Kubernetes cluster is halted, the fault is deleted, and the experiment as a whole is deleted. (CHAOS-782)
```

* Configurable response timeout for HTTP probes
    * This release adds a new response timeout parameter for HTTP probes in the user interface. The response timeout is in units of seconds. You can use this parameter to specify timeouts during HTTP probe health checks during chaos fault execution.

### Early access features

No early access features are available in this release.

### Fixed issues
* Enterprise ChaosHub appeared in the search results irrespective of the terms searched. Now it has been fixed.
* Details of a previously connected chaos infrastructure were prefilled when connecting to a new chaos infrastructure. Now it has been fixed. (CHAOS-777)
* The **Run** button was activated even when the chaos experiment was running. Now, the button is reactivated only after the chaos experiment is complete.(CHAOS-807)
* The chaos access page shows all experiments and experiment runs instead of showing experiments that were performed within a specific time frame. (CHAOS-810, CHAOS-762) 
* A cancel button and a back button have been added to the enable chaos screen. The buttons have made it easy to navigate between screens when setting up the chaos infrastructure.

> * When you search for a specific chaos fault and the chaos manager cannot map this chaos fault to a chaos fault icon, the user interface used to display an error. Now, instead of showing the error, it silently skips the error logs. (CHAOS-814)

Reviewer 3: ```suggestion
* When you search for a specific chaos fault and the chaos manager cannot map this chaos fault to a chaos fault icon, the user interface previously displayed an error. Now, instead of showing the error, it silently skips the error logs. (CHAOS-814)
```

* There are no early access features in this release. 

### Fixed issues

> * When a component on the user interface was missing due to incompatibilities, the page would crash. Now it has been fixed so that instead of the page crashing, the component field shows as empty. (Chaos-843)

Reviewer 2: Determine whether you can use "stops responding" instead of "crash" wherever you have to use the latter. It's less jarring on the imagination. Some style guides prohibit its usage. See https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/crash. (Might need to search and replace with appropriate text globally.)
